http/parse.go

```diff
@@ -211,10 +212,10 @@ func parseResponse(httpRes *http.Response, req *cmds.Request) (cmds.Response, er
 			e.Message = string(mes)
 			e.Code = cmdkit.ErrNormal
 		default:
-			// handle marshalled errors
-			err := res.dec.Decode(&e)
-			if err != nil {
-				return nil, err
+			// handle errors from headers
+			errStr := httpRes.Header.Get(StreamErrHeader)
```

Comment:
Is this correct? I think we want to set `res.initErr` and only *return* an error if we have some kind of parse error, right?

---

Also, we should probably maintain compatibility with existing servers (for now). We should probably.

1. Check the header first.
2. Check the stream next.

Speaking of which, we may want to do the inverse: Always send errors as values along with the header (just in case). Is that possible?

---

Basically, I'm worried about the case where a user calls `ipfs --api=/some/remote/server ...`.

Reply:
> Is this correct? I think we want to set res.initErr and only return an error if we have some kind of parse error, right?

Well, `parseResponse` is called by `Send` which is called by `Execute`, so the difference is whether you receive the error from `Send`/`Execute` or from `Next()`. IMHO returning the error from `Send` is fine, and making `Next` return the error would make the code more complex because you would have to make a noop `Response` that only returns an error and make `parseResponse` return that. I mean we can totally do that but I don't see the benefit.

> Also, we should probably maintain compatibility with existing servers (for now). We should probably...

IIUC the header is only set on connection close so I'd only check it when we receive an EOF. Does that sound reasonable to you?

> Speaking of which, we may want to do the inverse: Always send errors as values along with the header (just in case). Is that possible?

I think since the other side always checks both, it's okay to only set the header, right?

> Basically, I'm worried about the case where a user calls ipfs --api=/some/remote/server ....

Hmm, I don't get what's so special about that one. Can you elaborate?

Reply:
Oh I see what you mean now. Yes setting `res.initErr` seems to be the right thing to do.

Reply:
> IIUC the header is only set on connection close so I'd only check it when we receive an EOF. Does that sound reasonable to you?

Yes.

> I think since the other side always checks both, it's okay to only set the header, right?

Is that *currently* true?

> Hmm, I don't get what's so special about that one. Can you elaborate?

The client version may not match the server version. If the current client doesn't always read the header or the current server version doesn't always send it, bad things may happen.

Reply:
Hmm, I'm afraid you are right and we'll have to send errors also as values. Damn, I hoped we would not have to do that anymore...